TemperatureMeter.c

```diff
@@ -0,0 +1,154 @@
+/*
+htop - TemperatureMeter.c
+(C) 2013 Ralf Stemmer
+(C) 2014 Blair Bonnett
+(C) 2020 Maxim Kurnosenko
+Released under the GNU GPL, see the COPYING file
+in the source distribution for its full text.
+*/
+
+#include "TemperatureMeter.h"
+
+#include "ProcessList.h"
+#include "CRT.h"
+#include "StringUtils.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <limits.h>
+#include <errno.h>
+#include <sensors/sensors.h>
+
+static int get_cpu_sensors_amount(void) {
+   int n = 0, count = 0;
+
+   for (const sensors_chip_name *chip = sensors_get_detected_chips(NULL, &n); chip; chip = sensors_get_detected_chips(NULL, &n)) {
+      char buffer[64];
+      if (sensors_snprintf_chip_name(buffer, sizeof(buffer), chip) < 0)
+         continue;
+
+      // cpu_thermal is for Raspberry Pi sensor support
+      if ((String_startsWith(buffer, "coretemp") == 0) && (String_startsWith(buffer, "cpu_thermal") == 0))
+         continue;
+
+      int m = 0;
+      for (const sensors_feature *feature = sensors_get_features(chip, &m); feature; feature = sensors_get_features(chip, &m)) {
+         if (feature->type == SENSORS_FEATURE_TEMP)
+            count++;
+      }
+   }
+
+   return count;
+}
+
+static int get_cpu_temp(Meter* this, double *temp, int core, sensors_subfeature_type type) {
+   int n = 0;
+
+   for (const sensors_chip_name *chip = sensors_get_detected_chips(NULL, &n); chip; chip = sensors_get_detected_chips(NULL, &n)) {
+      char buffer[64];
+      if (sensors_snprintf_chip_name(buffer, sizeof(buffer), chip) < 0)
+         continue;
+
+      // cpu_thermal is for Raspberry Pi sensor support
+      if ((String_startsWith(buffer, "coretemp") == 0) && (String_startsWith(buffer, "cpu_thermal") == 0))
+         continue;
+
+      int m = 0;
+      for (const sensors_feature *feature = sensors_get_features(chip, &m); feature; feature = sensors_get_features(chip, &m)) {
+         if (feature->type != SENSORS_FEATURE_TEMP)
+            continue;
+
+         if (feature->number != core)
+            continue;
+
+         const sensors_subfeature *sub_feature = sensors_get_subfeature(chip, feature, type);
+         if (sub_feature) {
+            double local_temp;
+            if (sensors_get_value(chip, sub_feature->number, &local_temp) == 0) {
+               *temp = this->pl->settings->fahrenheitTemperature ? local_temp * 9 / 5 + 32 : local_temp;
```

Comment:
More some nitpicking, but the Fahrenheit conversion should be a separate function to decouple the value acquisition and display conversion logically.